### PR TITLE
feat(examples): add SHA-256 cryptographic hash implementation in Cairo-M

### PR DESCRIPTION
**Pull request by obatirou**
Close CORE-1135

Implement SHA-256 hash algorithm with:
- Complete hash algorithm with message scheduling and compression
- Support for pre-padded messages as u32 arrays
- Comprehensive Rust integration tests validating against reference vectors
- Project setup with cairom.toml manifest and build configuration

Limitations:
- Requires pre-padded input (no automatic padding)
- Fixed-size message arrays only

### CODE REVIEW

File: examples/sha256-cairo-m/src/sha256.cm [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    let bits_32: u32 = 32;
+    let complement_n: u32 = bits_32 - n;
+    let pow2_complement = pow2(complement_n);
+    let shr_part = x / pow2_n;
+    let shl_part = x * pow2_complement;
+    return shr_part | shl_part;
+}
+
+/// Computes 2^n for n < 32. Returns 0 for n >= 32.
+fn pow2(n: u32) -> u32 {
```
Comment (enitrat) [inline examples/sha256-cairo-m/src/sha256.cm:17]: use constant table

File: examples/sha256-cairo-m/src/sha256.cm [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    ];
+    // Message schedule array
+    let W: [u32; 64] = [
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+    ];
+    
+    // Copy chunk into first 16 words of message schedule
```
Comment (enitrat) [inline examples/sha256-cairo-m/src/sha256.cm:83]: fyi: we have for loops

File: examples/sha256-cairo-m/tests/integration_test.rs [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+#[cfg(test)]
+mod helpers {
+    use super::*;
+
+    /// SHA-256 rotate right operation
+    #[inline]
+    fn rust_rotr(x: u32, n: u32) -> u32 {
+        x.rotate_right(n)
+    }
```
Comment (enitrat) [inline examples/sha256-cairo-m/tests/integration_test.rs:300]:
```suggestion
```

File: examples/sha256-cairo-m/tests/integration_test.rs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+            InputValue::Number(z as i64),
+        ];
+        assert_cairo_u32_result!(&COMPILED_PROGRAM, function_name, args, expected);
+    }
+
+    proptest! {
+        #[test]
+        fn test_rotr(x: u32, n in 0u32..32) {
+            let args = vec![InputValue::Number(x as i64), InputValue::Number(n as i64)];
+            let expected = rust_rotr(x, n);
```
Comment (enitrat) [inline examples/sha256-cairo-m/tests/integration_test.rs:358]: x.rotate_right(n)

File: examples/sha256-cairo-m/src/sha256.cm [hunk 5]
```diff
@@ -0,0 +1,134 @@
+// SHA-256 cryptographic hash function implementation in Cairo-M.
+// This implementation follows NIST FIPS 180-4 specification.
+// It operates on pre-padded messages represented as arrays of u32 words.
+/// Rotate right operation: rotates bits of x right by n positions.
+fn rotr(x: u32, n: u32) -> u32 {
+    if n >= 32 {
+        return x;
+    } // Pre-computed powers of 2 for n = 0 to 31
+    let POW2: [u32; 32] = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304, 8388608, 16777216, 33554432, 67108864, 134217728, 268435456, 536870912, 1073741824, 2147483648];
```
[the remark this answers is not in the document]
Comment (obatirou) [reply examples/sha256-cairo-m/src/sha256.cm:9]: ooh nice, did not know
Will do that

File: examples/sha256-cairo-m/src/sha256.cm [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        2600822924,
+        528734635,
+        1541459225
+    ];
+    
+    let chunk_index = 0;
+    while (chunk_index != num_chunks) {
+        let current_chunk: [u32; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+        let i = 0;
+        while (i != 16) {
```
[the remark this answers is not in the document]
Comment (enitrat) [reply examples/sha256-cairo-m/src/sha256.cm:132]: > from my understanding: for loops need to use u32 and not felt.

`for (let i = 0; i != 16; i ++)` - `for (<init>; <cond>; <step>)`, any valid syntax at any other place would be valid there, unless I am missing something myself

File: examples/sha256-cairo-m/README.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+- Uses 32-bit word operations compatible with Cairo-M's u32 type
+- Implements all required bitwise operations (rotate, shift, XOR, AND)
+- Supports messages up to 2 chunks (128 bytes) in the current configuration
+
+### Cairo-M Features Used
+
+- **u32 arithmetic**: Native support for 32-bit unsigned integer operations
+- **Arrays**: Fixed-size arrays for message buffers and hash state
+- **Control flow**: While loops for chunk processing and message scheduling
+- **Bitwise operations**: XOR, AND, and custom rotate-right implementation
```
Comment (enitrat) [inline examples/sha256-cairo-m/README.md:64]:
```suggestion
```

File: examples/sha256-cairo-m/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+## Implementation Details
+
+### SHA-256 Algorithm
+
+This implementation follows the NIST FIPS 180-4 specification for SHA-256:
+
+- Processes messages in 512-bit (64-byte) chunks
+- Uses 32-bit word operations compatible with Cairo-M's u32 type
+- Implements all required bitwise operations (rotate, shift, XOR, AND)
+- Supports messages up to 2 chunks (128 bytes) in the current configuration
```
Comment (enitrat) [inline examples/sha256-cairo-m/README.md:57]: I think the real useful info here is our limitation on pre-padding and the size of input messages; can we make it clear somewhere?

File: cspell.yaml [hunk 9]
```diff
@@ -40,6 +40,7 @@ words:
   - extractfield
   - extracttuple
   - filetime
+  - FIPS
```
Comment (ClementWalter) [inline cspell.yaml:43]: keep all lower cased

File: examples/sha256-cairo-m/tests/integration_test.rs [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        test_sha256(&input);
+    }
+}
+
+// ==================================================================================
+// UNIT TESTS FOR HELPER FUNCTIONS
+// ==================================================================================
+
+#[cfg(test)]
+mod helpers {
```
Comment (ClementWalter) [inline examples/sha256-cairo-m/tests/integration_test.rs:302]: do we really need these `rust_` prefix ?

File: examples/sha256-cairo-m/src/sha256.cm [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+// SHA-256 round constants (first 32 bits of fractional parts of cube roots of first 64 primes)
+const K = [1116352408u32, 1899447441, 3049323471, 3921009573, 961987163, 1508970993, 2453635748, 2870763221, 3624381080, 310598401, 607225278, 1426881987, 1925078388, 2162078206, 2614888103, 3248222580, 3835390401, 4022224774, 264347078, 604807628, 770255983, 1249150122, 1555081692, 1996064986, 2554220882, 2821834349, 2952996808, 3210313671, 3336571891, 3584528711, 113926993, 338241895, 666307205, 773529912, 1294757372, 1396182291, 1695183700, 1986661051, 2177026350, 2456956037, 2730485921, 2820302411, 3259730800, 3345764771, 3516065817, 3600352804, 4094571909, 275423344, 430227734, 506948616, 659060556, 883997877, 958139571, 1322822218, 1537002063, 1747873779, 1955562222, 2024104815, 2227730452, 2361852424, 2428436474, 2756734187, 3204031479, 3329325298];
+
+// Maximum value for 32-bit unsigned integer
+const U32_MAX = 4294967295u32;
+
+/// Rotate right operation: rotates bits of x right by n positions.
+fn rotr(x: u32, n: u32) -> u32 {
```
[the remark this answers is not in the document]
Comment (obatirou) [reply examples/sha256-cairo-m/src/sha256.cm:15]: there is no conversion from felt to u32

File: examples/sha256-cairo-m/src/sha256.cm [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    if n == 0 {
+        return x;
+    }
+    let n_felt: felt = n as felt;
+    let pow2_n = POW2[n_felt];
+    let complement_n = 32 - n_felt;
+    let pow2_complement = POW2[complement_n];
+    let shr_part = x / pow2_n;
+    let shl_part = x * pow2_complement;
+    return shr_part | shl_part;
```
Comment (ClementWalter) [inline examples/sha256-cairo-m/src/sha256.cm:26]: this really got me confused!

```suggestion
    return shl_part | shr_part;
```

File: examples/sha256-cairo-m/README.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+- **Integration tests**: Comprehensive test vectors including edge cases
+- **Reference comparison**: All results verified against Rust's `sha2` crate
+
+## Extending the Implementation
+
+To modify the maximum message size:
+
+1. Update `MAX_CHUNKS` in `integration_test.rs`
+2. Adjust `PADDED_BUFFER_U32_SIZE` accordingly
+3. Update the fixed array size in `sha256.cm` to match
```
Comment (enitrat) [inline examples/sha256-cairo-m/README.md:81]: I dont really understand why I would want to update the message size nor why i have to do this - if this is a limit of our impl it should be clearly stated
